File: src/lib/forward.ts
import { attachmentsToForward } from "./kobo.js";
import type { KoboSubmission } from "./kobo.js";
import { transcribeAudio } from "./transcribe.js";
import { extractFields } from "./extract.js";
import { analyzeAudioText } from "./analyzeAudio.js";
import { extractTextFields } from "./extractText.js";
import { geocodeAddress } from "./geocode.js";
import type { EnrichmentStepResult } from "../types.js";

const EU_HOSTNAME = "eu.kobotoolbox.org";

type PromptField = { key: string; instruction: string; geocode?: boolean };
type PromptEntry = { description?: string; fields: PromptField[] };
type PromptMap = Record<string, PromptEntry>;

/**
 * Converts structured output-field definitions into a plain-text prompt for the LLM.
 * Prepends an optional global description/context before the field list.
 */
function buildPromptFromFields(stored: PromptEntry): string {
  const parts: string[] = [];
  if (stored.description?.trim()) parts.push(stored.description.trim());
  const lines = stored.fields
    .filter((f) => f.key.trim())
    .map((f) => (f.instruction.trim() ? `- ${f.key}: ${f.instruction}` : `- ${f.key}`));
  if (lines.length > 0) {
    parts.push(
      "Extract the following fields and return them as a JSON object with exactly these keys:\n" +
      lines.join("\n")
    );
  }
  return parts.join("\n\n");
}

/**
 * For each extracted field flagged `geocode: true` in the prompt config, forward-
 * geocode its string value via `geocodeAddress()` and write the resulting
 * `_latitude` / `_longitude` / `_adm{n}_*` fields into both `payload` and
 * `enrichment`, prefixed with the extracted field's key
 * (e.g. `address` → `address_adm1_name`).
 *
 * Returns the list of keys written, so the caller can append them to the step's
 * `keys` array for logging. Never throws — `geocodeAddress()` swallows its errors.
 */
async function geocodeExtractedFields(
  prompts: PromptMap | undefined,
  questionName: string,
  extracted: Record<string, string>,
  payload: Record<string, unknown>,
  enrichment: Record<string, string>
): Promise<string[]> {
  const geocodeKeys = (prompts?.[questionName]?.fields ?? [])
    .filter((f) => f.geocode && f.key.trim())
    .map((f) => f.key.trim());
  if (geocodeKeys.length === 0) return [];

  const written: string[] = [];
  await Promise.all(
    geocodeKeys.map(async (key) => {
      const value = extracted[key];
      if (typeof value !== "string" || !value.trim()) return;
      try {
        const geo = await geocodeAddress(value.trim());
        for (const [k, v] of Object.entries(geo)) {
          const prefixedKey = `${key}${k}`;
          payload[prefixedKey] = v;
          enrichment[prefixedKey] = v;
          written.push(prefixedKey);
        }
      } catch (err) {
        console.error(`[geocode-extracted] Error geocoding "${key}" for "${questionName}":`, err);
      }
    })
  );
  return written;
}

export interface ForwardResult {
  ok: boolean;
  httpStatus?: number;
  responseBody?: string;
  error?: string;
  /** Keys written into the payload during enrichment (transcripts + descriptions). Used by the edit-back step. */
  enrichment?: Record<string, string>;
  steps?: {
    transcribe?: Record<string, EnrichmentStepResult>;
    analyzeAudio?: Record<string, EnrichmentStepResult>;
    extract?: Record<string, EnrichmentStepResult>;
    extractText?: Record<string, EnrichmentStepResult>;
  };
}

function resolveKoboToken(
  koboBaseUrl: string,
  tokens: { global: string; eu: string }
): string {
  try {
    const host = new URL(koboBaseUrl).hostname;
    return host === EU_HOSTNAME ? tokens.eu : tokens.global;
  } catch {
    return tokens.global;
  }
}

/**
 * Fire-and-forget: fetches the images referenced in the submission and POSTs
 * everything to forwardUrl as multipart/form-data.
 *
 * Parts:
 *   - "submission"              — full submission JSON string (with transcripts injected)
 *   - "<media_file_basename>"   — one binary File part per referenced, non-deleted image
 *
 * If transcribeConfig and openaiApiKey are provided, audio attachments for each
 * named question are fetched and transcribed; the results are injected into the
 * submission JSON as "<questionName>_transcript" before forwarding.
 *
 * The correct wfp_logie Kobo token is selected based on koboBaseUrl hostname.
 * All errors are swallowed and logged — this function never throws.
 */
export async function forwardSubmission(
  submission: KoboSubmission,
  forwardUrl: string | undefined,
  koboBaseUrl: string,
  tokens: { global: string; eu: string },
  jsonPayload?: Record<string, unknown>,
  forwardToken?: string,
  transcribeConfig?: { questions: string[]; model?: string; prompt?: string; translateTo?: string },
  openaiApiKey?: string,
  forwardMedia?: string[],
  extractConfig?: { questions: string[]; model?: string; prompts?: PromptMap },
  analyzeAudioConfig?: { questions: string[]; model?: string; prompts?: PromptMap },
  extractTextConfig?: { questions: string[]; model?: string; prompts?: PromptMap },
  logieApiKey?: string
): Promise<ForwardResult> {
  try {
    const token = resolveKoboToken(koboBaseUrl, tokens);

    // Tracks enrichment values added to the payload — returned so the caller
    // can optionally write them back to Kobo via the edit-back step.
    const enrichment: Record<string, string> = {};
    const steps: NonNullable<ForwardResult["steps"]> = {};

    // Build the working payload; we may mutate it with transcript keys below.
    const payload: Record<string, unknown> = jsonPayload ? { ...jsonPayload } : { ...submission };

    // ── Transcription ──────────────────────────────────────────────────────
    if (transcribeConfig && openaiApiKey && transcribeConfig.questions.length > 0) {
      steps.transcribe = {};
      // Build a lookup map: question_xpath → attachment
      const audioByXpath = new Map(
        (submission._attachments ?? [])
          .filter((a) => !a.is_deleted && a.mimetype.startsWith("audio/"))
          .map((a) => [a.question_xpath, a])
      );

      await Promise.all(
        transcribeConfig.questions.map(async (questionName) => {
          const att = audioByXpath.get(questionName);
          if (!att) {
            console.warn(`[transcribe] No audio attachment found for question_xpath "${questionName}"`);
            steps.transcribe![questionName] = { ok: false, error: "No audio attachment found" };
            return;
          }
          try {
            const res = await fetch(att.download_url, {
              headers: { Authorization: `Token ${token}` },
            });
            if (!res.ok) {
              console.error(
                `[transcribe] Failed to fetch audio for "${questionName}": HTTP ${res.status}`
              );
              steps.transcribe![questionName] = { ok: false, error: `Failed to fetch audio: HTTP ${res.status}` };
              return;
            }
            const blob = await res.blob();
            const transcript = await transcribeAudio(
              blob,
              att.media_file_basename,
              openaiApiKey,
              transcribeConfig.model,
              transcribeConfig.prompt
            );
            if (transcript) {
              // Optionally translate the transcript into a target language
              let finalText = transcript;
              if (transcribeConfig.translateTo) {
                try {
                  const tlRes = await fetch("https://api.openai.com/v1/chat/completions", {
                    method: "POST",
                    headers: {
                      Authorization: `Bearer ${openaiApiKey}`,
                      "Content-Type": "application/json",
                    },
                    body: JSON.stringify({
                      model: "gpt-4o-mini",
                      messages: [
                        {
                          role: "system",
                          content: `Translate the following text to ${transcribeConfig.translateTo}. Return only the translated text, no explanation.`,
                        },
                        { role: "user", content: transcript },
                      ],
                      max_tokens: 1024,
                    }),
                  });
                  if (tlRes.ok) {
                    const tlData = await tlRes.json<{ choices?: Array<{ message?: { content?: string } }> }>();
                    finalText = tlData.choices?.[0]?.message?.content?.trim() ?? transcript;
                  } else {
                    console.error(`[transcribe] Translation failed for "${questionName}": HTTP ${tlRes.status}`);
                  }
                } catch (err) {
                  console.error(`[transcribe] Translation error for "${questionName}":`, err);
                }
              }
              const transcriptKey = `${questionName}_transcript`;
              payload[transcriptKey] = finalText;
              enrichment[transcriptKey] = finalText;
              steps.transcribe![questionName] = { ok: true, keys: [transcriptKey] };
            } else {
              steps.transcribe![questionName] = { ok: false, error: "No transcript returned" };
            }
          } catch (err) {
            console.error(`[transcribe] Error transcribing "${questionName}":`, err);
            steps.transcribe![questionName] = { ok: false, error: String(err) };
          }
        })
      );
    }
    // ── Audio analysis ─────────────────────────────────────────────────────
    if (analyzeAudioConfig && openaiApiKey && analyzeAudioConfig.questions.length > 0) {
      steps.analyzeAudio = {};
      const audioByXpathForAnalysis = new Map(
        (submission._attachments ?? [])
          .filter((a) => !a.is_deleted && a.mimetype.startsWith("audio/"))
          .map((a) => [a.question_xpath, a])
      );

      await Promise.all(
        analyzeAudioConfig.questions.map(async (questionName) => {
          try {
            // Reuse transcript already placed in payload by the transcription step
            let transcript = payload[`${questionName}_transcript`] as string | undefined;

            if (!transcript) {
              // Transcription not enabled for this question — transcribe fresh for analysis only
              const att = audioByXpathForAnalysis.get(questionName);
              if (!att) {
                console.warn(`[analyze-audio] No audio attachment found for question_xpath "${questionName}"`);
                steps.analyzeAudio![questionName] = { ok: false, error: "No audio attachment found" };
                return;
              }
              const res = await fetch(att.download_url, {
                headers: { Authorization: `Token ${token}` },
              });
              if (!res.ok) {
                console.error(`[analyze-audio] Failed to fetch audio for "${questionName}": HTTP ${res.status}`);
                steps.analyzeAudio![questionName] = { ok: false, error: `Failed to fetch audio: HTTP ${res.status}` };
                return;
              }
              const blob = await res.blob();
              transcript = await transcribeAudio(blob, att.media_file_basename, openaiApiKey) || undefined;
            }

            if (!transcript) {
              steps.analyzeAudio![questionName] = { ok: false, error: "No transcript available for analysis" };
              return;
            }

            const analyzeAudioFields = analyzeAudioConfig.prompts?.[questionName];
            const analyzed = await analyzeAudioText(
              transcript,
              openaiApiKey,
              analyzeAudioConfig.model,
              analyzeAudioFields && (analyzeAudioFields.description || analyzeAudioFields.fields?.length > 0)
                ? buildPromptFromFields(analyzeAudioFields)
                : undefined
            );
            if (analyzed) {
              const writtenKeys: string[] = [];
              for (const [k, v] of Object.entries(analyzed)) {
                if (k !== "_uuid") {
                  payload[k] = v;
                  enrichment[k] = v;
                  writtenKeys.push(k);
                }
              }
              const geoKeys = await geocodeExtractedFields(
                analyzeAudioConfig.prompts, questionName, analyzed, payload, enrichment
              );
              steps.analyzeAudio![questionName] = { ok: true, keys: [...writtenKeys, ...geoKeys] };
            } else {
              steps.analyzeAudio![questionName] = { ok: false, error: "No analysis returned" };
            }
          } catch (err) {
            console.error(`[analyze-audio] Error analyzing "${questionName}":`, err);
            steps.analyzeAudio![questionName] = { ok: false, error: String(err) };
          }
        })
      );
    }
    // ── Field extraction ───────────────────────────────────────────────────
    if (extractConfig && openaiApiKey && extractConfig.questions.length > 0) {
      steps.extract = {};
      const imageByXpath = new Map(
        (submission._attachments ?? [])
          .filter((a) => !a.is_deleted && a.mimetype.startsWith("image/"))
          .map((a) => [a.question_xpath, a])
      );

      await Promise.all(
        extractConfig.questions.map(async (questionName) => {
          const att = imageByXpath.get(questionName);
          if (!att) {
            console.warn(`[extract] No image attachment found for question_xpath "${questionName}"`);
            steps.extract![questionName] = { ok: false, error: "No image attachment found" };
            return;
          }
          try {
            const res = await fetch(att.download_url, {
              headers: { Authorization: `Token ${token}` },
            });
            if (!res.ok) {
              console.error(`[extract] Failed to fetch image for "${questionName}": HTTP ${res.status}`);
              steps.extract![questionName] = { ok: false, error: `Failed to fetch image: HTTP ${res.status}` };
              return;
            }
            const blob = await res.blob();
            const extractFields_ = extractConfig.prompts?.[questionName];
            const extracted = await extractFields(
              blob,
              att.media_file_basename,
              openaiApiKey,
              extractConfig.model,
              extractFields_ && (extractFields_.description || extractFields_.fields?.length > 0)
                ? buildPromptFromFields(extractFields_)
                : undefined
            );
            if (extracted) {
              const writtenKeys: string[] = [];
              for (const [k, v] of Object.entries(extracted)) {
                if (k !== "_uuid") {
                  payload[k] = v;
                  enrichment[k] = v;
                  writtenKeys.push(k);
                }
              }
              const geoKeys = await geocodeExtractedFields(
                extractConfig.prompts, questionName, extracted, payload, enrichment
              );
              steps.extract![questionName] = { ok: true, keys: [...writtenKeys, ...geoKeys] };
            } else {
              steps.extract![questionName] = { ok: false, error: "No fields extracted" };
            }
          } catch (err) {
            console.error(`[extract] Error extracting from "${questionName}":`, err);
            steps.extract![questionName] = { ok: false, error: String(err) };
          }
        })
      );
    }

    // ── Text field extraction ─────────────────────────────────────────────
    if (extractTextConfig && openaiApiKey && extractTextConfig.questions.length > 0) {
      steps.extractText = {};
      await Promise.all(
        extractTextConfig.questions.map(async (questionName) => {
          try {
            const text = (submission as Record<string, unknown>)[questionName];
            if (typeof text !== "string" || !text.trim()) {
              steps.extractText![questionName] = { ok: false, error: "No text value found for question" };
              return;
            }
            const extractTextFields_ = extractTextConfig.prompts?.[questionName];
            const extracted = await extractTextFields(
              text,
              openaiApiKey,
              extractTextConfig.model,
              extractTextFields_ && (extractTextFields_.description || extractTextFields_.fields?.length > 0)
                ? buildPromptFromFields(extractTextFields_)
                : undefined
            );
            if (extracted) {
              const writtenKeys: string[] = [];
              for (const [k, v] of Object.entries(extracted)) {
                if (k !== "_uuid") {
                  payload[k] = v;
                  enrichment[k] = v;
                  writtenKeys.push(k);
                }
              }
              const geoKeys = await geocodeExtractedFields(
                extractTextConfig.prompts, questionName, extracted, payload, enrichment
              );
              steps.extractText![questionName] = { ok: true, keys: [...writtenKeys, ...geoKeys] };
            } else {
              steps.extractText![questionName] = { ok: false, error: "No fields extracted" };
            }
          } catch (err) {
            console.error(`[extract-text] Error extracting from "${questionName}":`, err);
            steps.extractText![questionName] = { ok: false, error: String(err) };
          }
        })
      );
    }

    // If there is no forwarding target, return enrichment and stop here.
    if (!forwardUrl) {
      return { ok: true, enrichment, steps: Object.keys(steps).length > 0 ? steps : undefined };
    }

    // ── Attachment fetch & forward ─────────────────────────────────────────
    let attachments = attachmentsToForward(submission, jsonPayload);

    // Filter by allowed media types if the user has restricted forwarding
    if (forwardMedia && forwardMedia.length > 0) {
      attachments = attachments.filter((a) =>
        forwardMedia.some((prefix) => a.mimetype.startsWith(prefix + "/"))
      );
    }

    const form = new FormData();
    form.append("submission", JSON.stringify(payload));

    await Promise.all(
      attachments.map(async (att) => {
        try {
          const res = await fetch(att.download_url, {
            headers: { Authorization: `Token ${token}` },
          });
          if (!res.ok) {
            console.error(
              `[forward] Failed to fetch attachment ${att.media_file_basename}: HTTP ${res.status}`
            );
            return;
          }
          const blob = await res.blob();
          form.append(
            att.media_file_basename,
            new File([blob], att.media_file_basename, { type: att.mimetype })
          );
        } catch (err) {
          console.error(
            `[forward] Error fetching attachment ${att.media_file_basename}:`,
            err
          );
        }
      })
    );

    const fwdHeaders: HeadersInit = {};
    if (logieApiKey) {
      fwdHeaders["x-api-key"] = logieApiKey;
    } else if (forwardToken) {
      fwdHeaders["Authorization"] = `Bearer ${forwardToken}`;
    }

    const fwdRes = await fetch(forwardUrl, {
      method: "POST",
      headers: fwdHeaders,
      body: form,
    });

    const responseBody = await fwdRes.text().catch(() => undefined);
    const truncatedBody = responseBody ? responseBody.slice(0, 2048) : undefined;
    const stepsResult = Object.keys(steps).length > 0 ? steps : undefined;
    if (!fwdRes.ok) {
      console.error(
        `[forward] External service returned HTTP ${fwdRes.status} for ${forwardUrl}`
      );
      return { ok: false, httpStatus: fwdRes.status, responseBody: truncatedBody, enrichment, steps: stepsResult };
    }
    return { ok: true, httpStatus: fwdRes.status, responseBody: truncatedBody, enrichment, steps: stepsResult };
  } catch (err) {
    console.error("[forward] Unhandled error in forwardSubmission:", err);
    return { ok: false, error: String(err) };
  }
}

File: app/lib/forward.py
"""
Core enrichment + forwarding pipeline.
Port of src/lib/forward.ts.
"""

import logging

import httpx

from .kobo import attachments_to_forward
from .transcribe import transcribe_audio
from .extract import extract_fields
from .analyze_audio import analyze_audio_text
from .extract_text import extract_text_fields

logger = logging.getLogger(__name__)

EU_HOSTNAME = 'eu.kobotoolbox.org'


def _resolve_kobo_token(kobo_base_url: str, tokens: dict[str, str]) -> str:
    try:
        from urllib.parse import urlparse
        host = urlparse(kobo_base_url).hostname or ''
        return tokens['eu'] if host == EU_HOSTNAME else tokens['global']
    except Exception:
        return tokens['global']


def _build_prompt_from_fields(stored: dict) -> str:
    parts = []
    if stored.get('description', '').strip():
        parts.append(stored['description'].strip())
    lines = [
        f'- {f["key"]}: {f["instruction"]}' if f.get('instruction', '').strip() else f'- {f["key"]}'
        for f in stored.get('fields', [])
        if f.get('key', '').strip()
    ]
    if lines:
        parts.append(
            'Extract the following fields and return them as a JSON object with exactly these keys:\n' + '\n'.join(lines)
        )
    return '\n\n'.join(parts)


def forward_submission(
    submission: dict,
    forward_url: str | None,
    kobo_base_url: str,
    tokens: dict[str, str],
    json_payload: dict | None = None,
    forward_token: str | None = None,
    transcribe_config: dict | None = None,
    openai_api_key: str | None = None,
    forward_media: list[str] | None = None,
    extract_config: dict | None = None,
    analyze_audio_config: dict | None = None,
    extract_text_config: dict | None = None,
    logie_api_key: str | None = None,
) -> dict:
    """
    Runs enrichment (transcription, image extraction, audio analysis, text extraction)
    then optionally POSTs to forward_url as multipart/form-data.

    Returns a dict matching the ForwardResult shape from the TS code:
    {ok, http_status?, response_body?, error?, enrichment?, steps?}
    """
    try:
        token = _resolve_kobo_token(kobo_base_url, tokens)
        enrichment: dict[str, str] = {}
        steps: dict[str, dict] = {}

        payload: dict = dict(json_payload) if json_payload else dict(submission)

        # ── Transcription ──────────────────────────────────────────────────
        if transcribe_config and openai_api_key and transcribe_config.get('questions'):
            steps['transcribe'] = {}
            audio_by_xpath = {
                a['question_xpath']: a
                for a in (submission.get('_attachments') or [])
                if not a.get('is_deleted') and (a.get('mimetype') or '').startswith('audio/')
            }

            for question_name in transcribe_config['questions']:
                att = audio_by_xpath.get(question_name)
                if not att:
                    logger.warning('[transcribe] No audio attachment found for "%s"', question_name)
                    steps['transcribe'][question_name] = {'ok': False, 'error': 'No audio attachment found'}
                    continue
                try:
                    resp = httpx.get(
                        att['download_url'],
                        headers={'Authorization': f'Token {token}'},
                        timeout=30,
                    )
                    if not resp.is_success:
                        logger.error('[transcribe] Failed to fetch audio for "%s": HTTP %s', question_name, resp.status_code)
                        steps['transcribe'][question_name] = {'ok': False, 'error': f'Failed to fetch audio: HTTP {resp.status_code}'}
                        continue
                    transcript = transcribe_audio(
                        resp.content,
                        att['media_file_basename'],
                        openai_api_key,
                        transcribe_config.get('model', 'gpt-4o-mini-transcribe'),
                        transcribe_config.get('prompt'),
                    )
                    if transcript:
                        # Optional translation
                        final_text = transcript
                        translate_to = transcribe_config.get('translateTo')
                        if translate_to:
                            try:
                                tl_resp = httpx.post(
                                    'https://api.openai.com/v1/chat/completions',
                                    headers={'Authorization': f'Bearer {openai_api_key}', 'Content-Type': 'application/json'},
                                    json={
                                        'model': 'gpt-4o-mini',
                                        'messages': [
                                            {'role': 'system', 'content': f'Translate the following text to {translate_to}. Return only the translated text, no explanation.'},
                                            {'role': 'user', 'content': transcript},
                                        ],
                                        'max_tokens': 1024,
                                    },
                                    timeout=30,
                                )
                                if tl_resp.is_success:
                                    final_text = (tl_resp.json().get('choices') or [{}])[0].get('message', {}).get('content', '').strip() or transcript
                                else:
                                    logger.error('[transcribe] Translation failed for "%s": HTTP %s', question_name, tl_resp.status_code)
                            except Exception as exc:
                                logger.error('[transcribe] Translation error for "%s": %s', question_name, exc)

                        transcript_key = f'{question_name}_transcript'
                        payload[transcript_key] = final_text
                        enrichment[transcript_key] = final_text
                        steps['transcribe'][question_name] = {'ok': True, 'keys': [transcript_key]}
                    else:
                        steps['transcribe'][question_name] = {'ok': False, 'error': 'No transcript returned'}
                except Exception as exc:
                    logger.error('[transcribe] Error transcribing "%s": %s', question_name, exc)
                    steps['transcribe'][question_name] = {'ok': False, 'error': str(exc)}

        # ── Audio analysis ─────────────────────────────────────────────────
        if analyze_audio_config and openai_api_key and analyze_audio_config.get('questions'):
            steps['analyzeAudio'] = {}
            audio_by_xpath_analysis = {
                a['question_xpath']: a
                for a in (submission.get('_attachments') or [])
                if not a.get('is_deleted') and (a.get('mimetype') or '').startswith('audio/')
            }

            for question_name in analyze_audio_config['questions']:
                try:
                    transcript = payload.get(f'{question_name}_transcript')
                    if not transcript:
                        att = audio_by_xpath_analysis.get(question_name)
                        if not att:
                            logger.warning('[analyze-audio] No audio attachment for "%s"', question_name)
                            steps['analyzeAudio'][question_name] = {'ok': False, 'error': 'No audio attachment found'}
                            continue
                        resp = httpx.get(
                            att['download_url'],
                            headers={'Authorization': f'Token {token}'},
                            timeout=30,
                        )
                        if not resp.is_success:
                            logger.error('[analyze-audio] Failed to fetch audio for "%s": HTTP %s', question_name, resp.status_code)
                            steps['analyzeAudio'][question_name] = {'ok': False, 'error': f'Failed to fetch audio: HTTP {resp.status_code}'}
                            continue
                        transcript = transcribe_audio(resp.content, att['media_file_basename'], openai_api_key) or None

                    if not transcript:
                        steps['analyzeAudio'][question_name] = {'ok': False, 'error': 'No transcript available for analysis'}
                        continue

                    prompt_entry = (analyze_audio_config.get('prompts') or {}).get(question_name)
                    custom_prompt = _build_prompt_from_fields(prompt_entry) if prompt_entry and (prompt_entry.get('description') or prompt_entry.get('fields')) else None

                    analyzed = analyze_audio_text(
                        transcript,
                        openai_api_key,
                        analyze_audio_config.get('model', 'gpt-4o-mini'),
                        custom_prompt or 'Analyze this transcript. Return a flat JSON object with string values only.',
                    )
                    if analyzed:
                        written_keys = []
                        for k, v in analyzed.items():
                            if k != '_uuid':
                                payload[k] = v
                                enrichment[k] = v
                                written_keys.append(k)
                        steps['analyzeAudio'][question_name] = {'ok': True, 'keys': written_keys}
                    else:
                        steps['analyzeAudio'][question_name] = {'ok': False, 'error': 'No analysis returned'}
                except Exception as exc:
                    logger.error('[analyze-audio] Error analyzing "%s": %s', question_name, exc)
                    steps['analyzeAudio'][question_name] = {'ok': False, 'error': str(exc)}

        # ── Image field extraction ─────────────────────────────────────────
        if extract_config and openai_api_key and extract_config.get('questions'):
            steps['extract'] = {}
            image_by_xpath = {
                a['question_xpath']: a
                for a in (submission.get('_attachments') or [])
                if not a.get('is_deleted') and (a.get('mimetype') or '').startswith('image/')
            }

            for question_name in extract_config['questions']:
                att = image_by_xpath.get(question_name)
                if not att:
                    logger.warning('[extract] No image attachment for "%s"', question_name)
                    steps['extract'][question_name] = {'ok': False, 'error': 'No image attachment found'}
                    continue
                try:
                    resp = httpx.get(
                        att['download_url'],
                        headers={'Authorization': f'Token {token}'},
                        timeout=30,
                    )
                    if not resp.is_success:
                        logger.error('[extract] Failed to fetch image for "%s": HTTP %s', question_name, resp.status_code)
                        steps['extract'][question_name] = {'ok': False, 'error': f'Failed to fetch image: HTTP {resp.status_code}'}
                        continue
                    prompt_entry = (extract_config.get('prompts') or {}).get(question_name)
                    custom_prompt = _build_prompt_from_fields(prompt_entry) if prompt_entry and (prompt_entry.get('description') or prompt_entry.get('fields')) else None

                    extracted = extract_fields(
                        resp.content,
                        att['media_file_basename'],
                        openai_api_key,
                        extract_config.get('model', 'gpt-4o-mini'),
                        custom_prompt or 'Extract structured data from this image. Return a flat JSON object with string values only.',
                    )
                    if extracted:
                        written_keys = []
                        for k, v in extracted.items():
                            if k != '_uuid':
                                payload[k] = v
                                enrichment[k] = v
                                written_keys.append(k)
                        steps['extract'][question_name] = {'ok': True, 'keys': written_keys}
                    else:
                        steps['extract'][question_name] = {'ok': False, 'error': 'No fields extracted'}
                except Exception as exc:
                    logger.error('[extract] Error extracting from "%s": %s', question_name, exc)
                    steps['extract'][question_name] = {'ok': False, 'error': str(exc)}

        # ── Text field extraction ──────────────────────────────────────────
        if extract_text_config and openai_api_key and extract_text_config.get('questions'):
            steps['extractText'] = {}
            for question_name in extract_text_config['questions']:
                try:
                    text = submission.get(question_name)
                    if not isinstance(text, str) or not text.strip():
                        steps['extractText'][question_name] = {'ok': False, 'error': 'No text value found for question'}
                        continue
                    prompt_entry = (extract_text_config.get('prompts') or {}).get(question_name)
                    custom_prompt = _build_prompt_from_fields(prompt_entry) if prompt_entry and (prompt_entry.get('description') or prompt_entry.get('fields')) else None

                    extracted = extract_text_fields(
                        text,
                        openai_api_key,
                        extract_text_config.get('model', 'gpt-4o-mini'),
                        custom_prompt or 'Extract named entities and key facts. Return a flat JSON object with string values only.',
                    )
                    if extracted:
                        written_keys = []
                        for k, v in extracted.items():
                            if k != '_uuid':
                                payload[k] = v
                                enrichment[k] = v
                                written_keys.append(k)
                        steps['extractText'][question_name] = {'ok': True, 'keys': written_keys}
                    else:
                        steps['extractText'][question_name] = {'ok': False, 'error': 'No fields extracted'}
                except Exception as exc:
                    logger.error('[extract-text] Error extracting from "%s": %s', question_name, exc)
                    steps['extractText'][question_name] = {'ok': False, 'error': str(exc)}

        steps_result = steps if steps else None

        if not forward_url:
            return {'ok': True, 'enrichment': enrichment, 'steps': steps_result}

        # ── Attachment fetch & multipart forward ───────────────────────────
        attachments = attachments_to_forward(submission, json_payload)
        if forward_media:
            attachments = [
                a for a in attachments
                if any((a.get('mimetype') or '').startswith(prefix + '/') for prefix in forward_media)
            ]

        import json as json_module
        files = [('submission', (None, json_module.dumps(payload), 'application/json'))]
        for att in attachments:
            try:
                resp = httpx.get(
                    att['download_url'],
                    headers={'Authorization': f'Token {token}'},
                    timeout=30,
                )
                if resp.is_success:
                    files.append((att['media_file_basename'], (att['media_file_basename'], resp.content, att.get('mimetype', 'application/octet-stream'))))
                else:
                    logger.error('[forward] Failed to fetch attachment %s: HTTP %s', att['media_file_basename'], resp.status_code)
            except Exception as exc:
                logger.error('[forward] Error fetching attachment %s: %s', att['media_file_basename'], exc)

        fwd_headers: dict[str, str] = {}
        if logie_api_key:
            fwd_headers['x-api-key'] = logie_api_key
        elif forward_token:
            fwd_headers['Authorization'] = f'Bearer {forward_token}'

        fwd_resp = httpx.post(forward_url, headers=fwd_headers, files=files, timeout=60)
        response_body = fwd_resp.text[:2048] if fwd_resp.text else None

        if not fwd_resp.is_success:
            logger.error('[forward] External service returned HTTP %s for %s', fwd_resp.status_code, forward_url)
            return {'ok': False, 'http_status': fwd_resp.status_code, 'response_body': response_body, 'enrichment': enrichment, 'steps': steps_result}

        return {'ok': True, 'http_status': fwd_resp.status_code, 'response_body': response_body, 'enrichment': enrichment, 'steps': steps_result}

    except Exception as exc:
        logger.error('[forward] Unhandled error in forward_submission: %s', exc)
        return {'ok': False, 'error': str(exc)}
